### ISSUE TITLE
Implement get_state(), load_state and clear_state correctly so we can restart hub without tearing down single user servers.

### ISSUE THREAD

**Issue by ociule** [ni/jupyterhub-winlocalprocessspawner#8]
Jupyter hub allows restarts (useful for reloading conf and clearing some faults) without tearing down the router and single-user servers. The cleanup_proxy and cleanup_serversconfig values are set to False to do this.

However winlocalprocessspawner does not implement load_state, get_state or clear_state. It's therefore reusing the parent's but that didn't work in my tests. Probably we need to keep a reference to the token.

Here's the parent's code: https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/spawner.py#L1311

Here's the relevant doc section: https://jupyterhub.readthedocs.io/en/stable/reference/spawners.html#spawner-state

I'm experimenting with keeping the state, trying to come up with a patch.

**Comment by adelcast** [comment]
Makes sense, this is a use case that haven't come up in my testing. Looking forward to a patch!

**Comment by ociule** [comment]
load_state, get_state or clear_state, as implemented in the LocalProcessSpawner parent class, are fine. It was LocalProcessSpawner.poll that was not fine (and not hiding it, there was a comment). It was using os.kill(0, pid) to check if a single user server with the given pid was running. This was failing on windows, as even signal 0 was killing the single user server. I've replaced that with some win specific code using kernel32 methods. Checking if a pid exists on windows was surprisingly tricky, as we need to cover processes owned by other users too for winlocalprocessspawner. The psutil package has a good implementation in a C extension, but ~~I do not wish to use psutil. Installing it has the side effect of enabling terminals in jupyter notebook - for this reason it is an optional requirement for notebook. I, for one, routinely remove psutil (see correction)~~.

I'm testing it a bit more and then I'll push it.

**Comment by ociule** [comment]
I can see two possible improvements:

 * currently, I reimplement LocalProcessSpawner.poll in WinLocalProcessSpawner, changing the single line `alive = await self._signal(0)` into `await = pid_exists(self.pid)`. It would be better to patch jupyterhub's LocalProcessSpawner.poll, changing the same line into `await = self.single_user_server_exists()` Then LocalProcessSpawner and WinLocalProcessSpawner can have platform specific implementations of single_user_server_exists.

 * LocalProcessSpawner.stop follows Unix best practices of sending an escalating sequence of SIGINT, SIGTERM and SIGKILL signals. On windows, this results in a brutal termination equivalent to sigkill from the first SIGINT attempt. Win best practice would be to send signal.CTRL_C_EVENT twice before SIGKILL, as the single-user server already expects this. I do not observe any bad effects from the current implementation on windows as yet.

I'm preparing an MR for jh to handle the first issue.

**Comment by adelcast** [comment]
Your approach at https://github.com/jupyterhub/jupyterhub/pull/2882 looks great to me. It should get merged into JupyterHub soon, as I think it fits nicely. Once it is merged, this PR can be updates, as you mentioned.

Regarding psutil, JH already relies on it to be able to kill the ConfigurarableHTTPProxy correctly (and its children). Check https://github.com/jupyterhub/jupyterhub/pull/1899.

If you want to disable terminals, you can use c.NotebookApp.terminals_enabled (c.NotebookApp.terminals_enabled). Will that work for your case?

**Comment by ociule** [comment]
Thanks for the tip, I will test that. It's weird that JH relies on psutil because it's not as of yet in their requirements.txt. I'm not sure how it gets installed, but indeed it gets every time I setup notebook / lab / hub ~~and I then usually remove it (see correction)~~.

**Comment by ociule** [comment]
Correction: it is not psutil that I usually remove but terminado. The interaction with terminals I described earlier does not apply to psutil, but to terminado.

So there's no issue whatsoever in using psutil for pid_exists. I will test it on my deployment.

**Comment by ociule** [comment]
If JH uses psutil, I should simplify my JH MR to use psutil.pid_exists directly.

**Comment by ociule** [comment]
All right, I've submitted my MR update to JH. There'll be some discussion on whether to make psutil a requirement on unix too (it's not currently). Depending on that I'll update my MR. After that, there's no patch to merge here: the poll in LocalProcessSpawner will work on win too.

**Comment by adelcast** [comment]
sounds good to me, thanks again for all the good work!

**Comment by ociule** [comment]
The poll in LocalProcessSpawner is now good to go on win32. 
We should update the doc to ask installing jupyterhub master (1.1.0b1 is missing this)
We can now close this.
